lib/datasource.js

```diff
@@ -2223,6 +2224,13 @@ DataSource.prototype.transaction = function(execute, options, cb) {
       // retrieved again in determineOptions() in dao.js, as well as referenced
       // in transaction.commit() and transaction.rollback() below.
       transaction.transaction = tx;
+      // Handle timeout and pass it on as an error.
+      tx.observe('timeout', function(context, next) {
+        const err = new Error(g.f('Transaction is rolled back due to timeout'));
+        err.code = 'TRANSACTION_TIMEOUT';
+        done(err);
+        next();
```

Comment:
I think the usual way is to pass errors through the `next` callback. Can we use it here too?

```js
tx.observe('timeout', function(context, next) {
  const err = new Error(g.f('Transaction is rolled back due to timeout'));
  err.code = 'TRANSACTION_TIMEOUT';
  next(err);
});
```

Reply:
Unfortunately that doesn't work as expected: Calling `next(err)` will trigger this default callback in `loopback-connect`: https://github.com/strongloop/loopback-connector/blob/c7e4f5856218724425bf05f0e69bc60f014071dc/lib/transaction.js#L110-L116

And as you can see, if an error is passed to it, the rollback isn't executed there.

So there's the choice between calling:

```js
next(err);
done(err);
```
OR

```js
next();
done(err);
```

OR

```js
done(err);
```

I can't tell right now which one is right, but I'm fairly certain that we should call `done(err)` because otherwise the promise will never resolve...

Reply:
At this point you know more about this part of our codebase than I do, so I am happy to follow your judgment. Maybe call `next` first, so that the transaction rollback is started before we return control back to our caller via `done`?

Reply:
I think this is the way to go:

```js
        next(err);
        done(err);
```

`Passing `err` to next()` prevents `tx.rollback()` from being called inside https://github.com/strongloop/loopback-connector/blob/c7e4f5856218724425bf05f0e69bc60f014071dc/lib/transaction.js#L110-L116 which is actually good, because we have no way of getting a callback for when it finished the rollback...

`done(err) handles the rollback in the code of this PR, and also rejects the promise with it at the end of the rollback, which is much more predictable behavior than having it execute in parallel after return.

Reply:
>  I'll add a code comment about it.

Yes please!